runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5278,6 +5287,46 @@ class INTERPRETER_CLASS
 			} else if (J9NtcPointer == argType) {
 				/* ffi_call expects the address of the pointer is the address of the stackslot. */
 				pointerValues[i] = (U_64)ffiArgs[i];
+#if JAVA_SPEC_VERSION >= 22
+				if ((U_64)J9_FFI_DOWNCALL_HEAP_ARGUMENT_ID == pointerValues[i]) {
+					j9object_t heapBase = (j9object_t)J9JAVAARRAYOFOBJECT_LOAD(
+							_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 11), /* The heap base array. */
+							curHeapArgIdx);
+					if (NULL == heapBase) {
+						rc = THROW_NPE;
+						goto done;
+					}
+					U_64 heapOffset = (U_64)J9JAVAARRAYOFLONG_LOAD(
+							_currentThread,
+							J9_JNI_UNWRAP_REFERENCE(_sp + 10), /* The heap offset array. */
+							curHeapArgIdx);
+					/* Extract the heap address from the base object of the heap segment which is
+					 * created by MemorySegment.ofArray(), MemorySegment.ofBuffer() or other methods
+					 * (which only returns part of the heap array) in OpenJDK.
+					 *
+					 * Note:
+					 * 1) Only MemorySegment objects representing the on-heap arrays reach here.
+					 * 2) The heap base object must be a contiguous array to be accessed in native.
+					 */
+					if (J9ISCONTIGUOUSARRAY(_currentThread, heapBase)) {
+						/* The address is simply the base object plus the offset. */
+						pointerValues[i] = (UDATA)heapBase + heapOffset;
+						curHeapArgIdx += 1;
+					} else {
+						/* TODO: We will need to handle the case of a discontiguous array later
+						 * by copying the heap content to the native memory and copy it back
+						 * in the case of modifications in the heap array.
```

Comment:
I hope we don't have to resort to copying: it will kill performance that will be expected of "critical" down-calls.
I think it's enough to say we'll need to handle dis-contiguous arrays.